engine/src/main/java/org/hibernate/search/engine/backend/work/execution/OperationSubmitter.java

```diff
@@ -8,52 +8,128 @@
+	public static OperationSubmitter rejectedExecutionException() {
+		return REJECTED_EXECUTION_EXCEPTION;
+	}
+
+	/**
+	 * Creates an operation submitter that would attempt to submit work to a queue, but in case the queue is full it
+	 * would offload the submitting of the operation to provided executor.
+	 * This would never block the current thread, but the one to which the work is offloaded.
+	 *
+	 * @param executor executor to offload submit operation to if the queue is full
+	 */
+	public static OperationSubmitter offloading(Consumer<Runnable> executor) {
+		return new OffloadingExecutorOperationSubmitter( executor );
+	}
+
+	private static final class BlockingOperationSubmitter extends OperationSubmitter {
+		@Override
+		public <T> void submitToQueue(BlockingQueue<? super T> queue, T element,
+				Consumer<? super T> blockingRetryProducer)
+				throws InterruptedException {
+			queue.put( element );
+		}
+
 		@Override
-		public <T> void submitToQueue(BlockingQueue<T> queue, T element) {
+		public <T extends Runnable> void submitToExecutor(SimpleScheduledExecutor executor, T element,
+				Function<? super T, Runnable> blockingRetryProducer) {
+			executor.submit( element );
+		}
+	}
+
+	private static final class RejectedExecutionExceptionOperationSubmitter extends OperationSubmitter {
+		@Override
+		public <T> void submitToQueue(BlockingQueue<? super T> queue, T element,
+				Consumer<? super T> blockingRetryProducer) {
 			if ( !queue.offer( element ) ) {
 				throw new RejectedExecutionException();
 			}
 		}
-	};
 
-	/**
-	 * Defines how an element will be submitted to the queue. Currently supported implementations:
-	 * <ul>
-	 *     <li>{@link #BLOCKING}</li>
-	 *     <li>{@link #REJECTED_EXECUTION_EXCEPTION}</li>
-	 * </ul>
-	 * <p>
-	 * Depending on the implementation might throw {@link RejectedExecutionException}.
-	 *
-	 * @param queue
-	 * @param element
-	 */
-	public abstract <T> void submitToQueue(BlockingQueue<T> queue, T element) throws InterruptedException;
+		@Override
+		public <T extends Runnable> void submitToExecutor(SimpleScheduledExecutor executor, T element,
+				Function<? super T, Runnable> blockingRetryProducer) {
+			executor.offer( element );
+		}
+	}
+
+	private static final class OffloadingExecutorOperationSubmitter extends OperationSubmitter {
+		private final Consumer<Runnable> executor;
+
+		private OffloadingExecutorOperationSubmitter(Consumer<Runnable> executor) {
+			this.executor = executor;
+		}
+
+		@Override
+		public <T> void submitToQueue(BlockingQueue<? super T> queue, T element, Consumer<? super T> blockingRetryProducer) {
+			if ( !queue.offer( element ) ) {
+				executor.accept( () -> blockingRetryProducer.accept( element ) );
+			}
+		}
+
+		@Override
+		public <T extends Runnable> void submitToExecutor(SimpleScheduledExecutor executor, T element,
+				Function<? super T, Runnable> blockingRetryProducer) {
+			this.executor.accept( blockingRetryProducer.apply( element ) );
```

Comment:
I see you added an "offer" method to `SimpleScheduledExecutor`, but you're not using it here?

Reply:
right ... that `offer()` was mostly for the rejecting submitter (if we'd call `submitToExecutor` on it). In this case, I'm 50-50 as to if we should call it or not. On one side, yeah, we are trying not to block the thread, and if we are about to do so we are offloading the submit to another thread that would block, so it makes sense that if we cannot offload because that would block the current thread we should throw an exception. On the other hand, we are trying to offload - so wouldn't a user expect that it would actually get offloaded?

Reply:
Well I was thinking the `offloading` submitter should catch the `RejectedExecutionException` and offload?

Reply:
hmm so let's see ...

  | submit to queue | submit to executor
-- | -- | --
blocking | just submits to the queue and blocks | just submits to the executor and blocks
rejecting | check if the queue is full. If the queue still can store something - accept that work, if the queue is full - throw exception | that's where the new method is used for the executor. We don't want to block so we make an attempt to offer the new work to it. In that default implementation, it relies on `isBlocking()`. if the executor is blocking, we just fail, even if the executor is empty and can accept work without blocking
offloading | similar to rejecting attempts to add work to the queue, but if the queue is full, then makes "offload" to do blocking adding to the queue in some other thread. | just offloads to the other thread right away in a blocking manner to make sure that work is submitted at some point.

With that written down ... I think it makes sense to do:
```java
public <T extends Runnable> void submitToExecutor(SimpleScheduledExecutor executor, T element,
		Function<? super T, Runnable> blockingRetryProducer) {
	try {
		executor.offer( element );
	}
	catch (Exception e) {
		this.offloadingExecutor.accept( blockingRetryProducer.apply( element ) );
	}
}
```
was that what you had in mind?

Reply:
Well, that would be a `catch (RejectedOperationException e)`, because catching `Exception` is not what we want here. If `executor` is `null` and that leads to a `NullPointerException`, I'd rather not offload the operation, I want to propagate the `NullPointerException` :)

Actually, even catching `RejectedOperationException` is questionable (using exceptions for flow control being evil, and all that), which is why I initially suggested that the `offer` method return `null` when rejecting the operation, but I agree it's also dodgy in its own way. And, since rejecting an operation should be, well, exceptional, I guess `RejectedOperationException` makes sense. So, ok, `RejectedOperationException` it is.

But otherwise, yes, that's what I had in mind.

Reply:
Meh, not worth the trouble. Just keep `RejectedOperationException` :)